sips/sip-45.md

```diff
@@ -0,0 +1,78 @@
+
+| SIP-Number          | 45 |
+| ---:                | :--- |
+| Title               | Prioritized Transaction Submission |
+| Description         | Enable effective prioritized transaction submission. |
+| Author              | Shio Coder \<shio.coder@gmail.com\> |
+| Editor              | Will Riches \<will@sui.io, @wriches\>
+| Type                | Standard |
+| Category            | Core |
+| Created             | 2024-11-26 |
+| Comments-URI        | https://sips.sui.io/comments-45 |
+| Status              | Last Call |
+| Requires            | N/A |
+
+## Abstract
+
+Enable effective prioritized transaction submission based on gas price.
+
+## Motivation
+
+Currently, there is no deterministic gas ordering for the transactions, even when these transactions are submitted within an extremely short time window (less than 100ms). This is due to the fact that different transactions are randomly submitted by different validators, and the validator that is supposed to submit the transaction can have latency jitters. We propose a mechanism to mitigate the random factor and enable prioritized transaction submission.
```

Comment:
if I understand this correctly:
**Before SIP-45**
Today there is no way for a sender to adjust the gas price. If a validator reference price is 750 MIST, then the sender must pay that price.

**After SIP-45**
With SIP-45, the sender can now express a higher price. If a validator reference price is 750 MIST, then the sender can pay 750 MIST + a priority fee (e.g. 750 MIST additional) to bid for 'faster' inclusion?

______
This would imply that the validator reference price becomes a price floor as opposed to a strict price then right?

`Once a validator receives a transaction with a higher gas price, it is submitted immediately to consensus regardless of tx digest`

Is there any new mechanisms in the validator code that will force a validator to submit higher priority fee transactions faster? Or is it simply economic incentive from a higher number of values that allows the transaction to be "submitted immediately".

Reply:
Hi @Evan-Kim2028, thanks for taking a look at the proposal. What you described "after SIP-45" where the sender can specify any gas price and validators only admit those above the reference price are already the logic today. Ordering transactions within the same consensus commit by gas price before execution has also been implemented in Sui. These logic are unaffected by SIP-45.

What SIP-45 proposes to change are mainly these two area:
(1) increase max gas price for transactions, so post-consensus bidding (ordering) do not have their max bid (gas price) limited even when the opportunity is relatively profitable.
(2) use more consensus bandwidth to ensure txns paying higher gas price can more reliably be included in the next consensus commit. A new mechanism will be implemented for this. Sui has parallel block proposers, which can propose duplicated txns at the same time. Previously Sui leans a bit too much on the side of avoiding duplicated txns taking up consensus bandwidth. After SIP-45, Sui will allow txns paying higher gas price to use more consensus bandwidth to reduce the time variance of these txns getting committed.

To use the transportation metaphor, SIP-45 is like using N carriers instead of only one to try to deliver the same message, when the sender is paying N times the regular shipping fee. This improves the reliability of delivering the message as quickly as possible. But it does not affect the minimum time to deliver the message or how the message will be ordered later.

Hope this is useful!

Reply:
Thanks for the clarifications. I think the main changes make sense and I understand why the consensus bandwidth is expected to increase (although I don't think the analysis was particularly robust).

```A new mechanism will be implemented for this.```

Can you provide more details here? I was not able to find it in the SIP details. If I understand correctly at a high level, if I send the same transaction to N validators, then there is not a higher chance that the transaction gets certified faster to traverse the DAG to get majority certificates faster. This is the downside today.

How exactly will the new mechanism work in this multiple proposer setting to ensure more reliable inclusion as a transaction traverses through the DAG?

Reply:
Yes SIP usually focuses on the high level instead of its implementation. For details, you can first check out the logic in [consensus_adapter.rs](https://github.com/MystenLabs/sui/blob/8d0699ebee3bd6452e5f09084c9da85cd2e10adf/crates/sui-core/src/consensus_adapter.rs) for context. https://github.com/MystenLabs/sui/pull/20668 and https://github.com/MystenLabs/sui/pull/20738 are in progress changes to implement SIP-45. The intuition is that relying on a single validator to submit a transaction to consensus can be inconsistent because of networking and other faults, so using multiple validators for consensus submissions at the same time can improve the submission reliability. If you have more questions on background / context, we can follow up on X ([@mwtian](https://x.com/mwtian)).